aria/modeling/service.py

```diff
@@ -0,0 +1,1529 @@
+    :ivar scaling_groups: ??
+    :vartype scaling_groups: {}
+
+    :ivar modifications: Modifications of this service
+    :vartype modifications: [:class:`ServiceModification`]
+    :ivar updates: Updates of this service
+    :vartype updates: [:class:`ServiceUpdate`]
+    :ivar executions: Executions on this service
+    :vartype executions: [:class:`Execution`]
+    """
+
+    __tablename__ = 'service'
+
+    @declared_attr
+    def service_template(cls):
+        return cls.many_to_one_relationship('service_template')
+
+    description = Column(Text)
+
+    @declared_attr
+    def meta_data(cls):
+        # Warning! We cannot use the attr name "metadata" because it's used by SqlAlchemy!
+        return cls.many_to_many_relationship('metadata', dict_key='name')
+
+    @declared_attr
+    def nodes(cls):
+        return cls.one_to_many_relationship('node')
+
+    @declared_attr
+    def groups(cls):
+        return cls.one_to_many_relationship('group')
+
+    @declared_attr
+    def policies(cls):
+        return cls.one_to_many_relationship('policy')
+
+    @declared_attr
+    def substitution(cls):
+        return cls.one_to_one_relationship('substitution')
+
+    @declared_attr
+    def inputs(cls):
+        return cls.many_to_many_relationship('parameter', table_prefix='inputs',
+                                             dict_key='name')
+
+    @declared_attr
+    def outputs(cls):
+        return cls.many_to_many_relationship('parameter', table_prefix='outputs',
+                                             dict_key='name')
+
+    @declared_attr
+    def operations(cls):
+        return cls.one_to_many_relationship('operation', dict_key='name')
+
+    @declared_attr
+    def plugins(cls):
+        return cls.many_to_many_relationship('plugin')
+
+    created_at = Column(DateTime, nullable=False, index=True)
+    updated_at = Column(DateTime)
+
+    # region orchestration
+
+    permalink = Column(Text)
+    scaling_groups = Column(modeling_types.Dict)
+
+    # endregion
+
+    # region foreign keys
+
+    __private_fields__ = ['substituion_fk',
+                          'service_template_fk']
+
+    # Service one-to-one to Substitution
+    @declared_attr
+    def substitution_fk(cls):
+        return cls.foreign_key('substitution', nullable=True)
+
+    # Service many-to-one to ServiceTemplate
+    @declared_attr
+    def service_template_fk(cls):
+        return cls.foreign_key('service_template', nullable=True)
+
+    # endregion
+
+    def satisfy_requirements(self, context):
+        satisfied = True
+        for node in self.nodes:
+            if not node.satisfy_requirements(context):
+                satisfied = False
+        return satisfied
+
+    def validate_capabilities(self, context):
+        satisfied = True
+        for node in self.nodes:
+            if not node.validate_capabilities(context):
+                satisfied = False
+        return satisfied
+
+    def find_nodes(self, node_template_name):
```

Comment:
couldn't this be replaced by an sql query (in a single line), and if so, why do we need this helper method?